File: app/tools/chatwoot_tools.py
"""Ferramentas de integração com Chatwoot utilizadas pelo Koalla."""
from langchain_core.tools import tool
from typing import Annotated
from app.services import chatwoot_client


# These tools receive context via closure — set before agent invocation
_ctx: dict = {}


def set_context(ctx: dict) -> None:
    """Inject conversation context (account_id, conversation_id, etc.)."""
    _ctx.update(ctx)


@tool
async def send_text(
    content: Annotated[str, "The text message to send to the user."]
) -> str:
    """
    Send a text message to the user via Chatwoot.
    Use for links, phone numbers, or formatted info that doesn't work well as audio.
    NEVER include the sent data again in your output after calling this tool.
    """
    await chatwoot_client.send_message(
        _ctx["account_id"], _ctx["conversation_id"], content
    )
    return "Message sent."


@tool
async def react_to_message(
    emoji: Annotated[str, "The emoji to react with. Allowed: 😀 ❤️ 👍 👀 ✅"]
) -> str:
    """
    Send an emoji reaction to the user's last message.
    Use max 3 times per conversation.
    NEVER use this tool multiple times in a row.
    """
    await chatwoot_client.send_reaction(
        _ctx["account_id"],
        _ctx["conversation_id"],
        int(_ctx["message_id"]),
        emoji,
    )
    return "Reaction sent."


@tool
async def set_response_preference(
    preference: Annotated[str, "User preference: 'audio' or 'texto'"]
) -> str:
    """
    Save the user's preference for receiving responses as audio or text.
    Only use when the user explicitly says they want only audio or only text.
    """
    current_attrs = _ctx.get("contact_custom_attributes", {})
    await chatwoot_client.update_contact_attributes(
        _ctx["account_id"],
        _ctx["contact_id"],
        {**current_attrs, "preferencia_audio_texto": preference},
    )
    return f"Preference set to '{preference}'."


@tool
async def send_cancellation_alert(
    message: Annotated[str, "Alert message with name, date/time, reason and notes."]
) -> str:
    """Send a cancellation alert to the manager's conversation."""
    alert_conv_id = _ctx.get("alert_conversation_id")
    if not alert_conv_id:
        return "Alert conversation not configured."
    await chatwoot_client.send_message(_ctx["account_id"], int(alert_conv_id), message)
    return "Alert sent."

File: src/main/kotlin/ai/koalla/core/tools/ChatwootTools.kt
package ai.koalla.core.tools

import ai.koalla.core.gateway.ChatwootGateway
import kotlinx.coroutines.runBlocking
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.Description
import java.util.function.Function

/**
 * Spring AI function beans for Chatwoot messaging operations.
 */
@Configuration
class ChatwootTools(
    private val contextHolder: ToolContextHolder,
    private val chatwootGateway: ChatwootGateway
) {

    data class SendTextRequest(val content: String = "")

    @Bean("sendText")
    @Description("""
        Send a text message to the user via Chatwoot.
        Use for links, phone numbers, PIX keys, or formatted info.
        NEVER include the sent data again in your output after calling this tool.
        Parameters:
        - content: The text message to send
    """)
    fun sendText(): Function<SendTextRequest, String> = Function { req ->
        val ctx = contextHolder.require()
        runBlocking {
            chatwootGateway.sendMessage(ctx.accountId, ctx.conversationId, req.content)
        }
        "Message sent."
    }

    data class ReactToMessageRequest(val emoji: String = "👍")

    @Bean("reactToMessage")
    @Description("""
        Send an emoji reaction to the user's last message.
        Use max 3 times per conversation. NEVER use multiple times in a row.
        Parameters:
        - emoji: Allowed: 😀 ❤️ 👍 👀 ✅
    """)
    fun reactToMessage(): Function<ReactToMessageRequest, String> = Function { req ->
        val ctx = contextHolder.require()
        val allowedEmojis = setOf("😀", "❤️", "👍", "👀", "✅")
        if (req.emoji !in allowedEmojis) {
            return@Function "Invalid emoji. Allowed: ${allowedEmojis.joinToString(" ")}"
        }
        runBlocking {
            chatwootGateway.sendReaction(ctx.accountId, ctx.conversationId, ctx.messageId, req.emoji)
        }
        "Reaction sent."
    }

    data class SetResponsePreferenceRequest(val preference: String = "texto")

    @Bean("setResponsePreference")
    @Description("""
        Save user's preference for audio or text responses.
        Only use when user explicitly requests.
        Parameters:
        - preference: 'audio' or 'texto'
    """)
    fun setResponsePreference(): Function<SetResponsePreferenceRequest, String> = Function { req ->
        val ctx = contextHolder.require()
        val validPrefs = setOf("audio", "texto")
        if (req.preference.lowercase() !in validPrefs) {
            return@Function "Invalid preference. Use 'audio' or 'texto'."
        }
        val currentAttrs = ctx.contactCustomAttributes.toMutableMap()
        currentAttrs["preferencia_audio_texto"] = req.preference.lowercase()
        runBlocking {
            chatwootGateway.updateContactAttributes(ctx.accountId, ctx.contactId, currentAttrs)
        }
        "Preference set to '${req.preference}'."
    }

    data class SendCancellationAlertRequest(val message: String = "")

    @Bean("sendCancellationAlert")
    @Description("""
        Send a cancellation alert to the manager's conversation.
        Parameters:
        - message: Alert with name, date/time, reason and notes
    """)
    fun sendCancellationAlert(): Function<SendCancellationAlertRequest, String> = Function { req ->
        val ctx = contextHolder.require()
        val alertConvId = ctx.alertConversationId
        if (alertConvId.isBlank()) {
            return@Function "Alert conversation not configured."
        }
        val alertConvIdInt = alertConvId.toIntOrNull()
            ?: return@Function "Invalid alert conversation ID."
        runBlocking {
            chatwootGateway.sendMessage(ctx.accountId, alertConvIdInt, req.message)
        }
        "Alert sent."
    }
}
